File: 2019/01/b.ex
defmodule AdventOfCode.Fuel do
  def calc(weight), do: Integer.floor_div(weight, 3) - 2
  def calc(weight, list) do
    fuel = calc(weight)
    if fuel > 0, do: calc(fuel, [fuel | list]), else: list
  end

  def exec(file) do
    File.read!(file)
    |> String.trim()
    |> String.split("\n")
    |> Enum.map(&String.to_integer/1)
    |> Enum.map(&calc(&1, []))
    |> List.flatten()
    |> Enum.sum()
  end
end

File: 2019/01/b.rb
def calc_fuel(weight, list = [])
  fuel = calc_fuel_basic(weight)
  if fuel > 0
    list << fuel
    calc_fuel(fuel, list)
  else
    return list
  end
end

def calc_fuel_basic(weight)
  weight / 3 - 2
end

lines = File.read("data.txt").split("\n")
puts lines.map { |w| calc_fuel(w.to_i) }.flatten.sum
